daemon: Merge Envoy logs with cilium logs by default.

jrajahalme opened the pull request:
Use a file logger for Envoy only if non-empty '--envoy-log' command
line parameter is given. Otherwise merge Envoy logs with Cilium logs
with 'subsys=envoy-<part>', where '<part>' is Envoy's logger name
(e.g., "filter", "upstream", "router", etc.).

Envoy rebase is needed to get support for the new '--log-format' command
line option.

jrajahalme committed:
envoy: Rebase to Envoy master.

Need support for `--log-format` command line option.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

rlenglet commented on envoy/WORKSPACE:
```diff
@@ -7,7 +7,7 @@ workspace(name = "cilium")
 # No other line in this file may have ENVOY_SHA followed by an equals sign!
 #
-ENVOY_SHA = "12434bee6a6be56a62955999ed38e16a7a9ee054"
+ENVOY_SHA = "fc3693b5205518171a590d87e51ae317040f9276"
```
Please pull the changes on the Envoy side out of this PR. Those have nothing to do with logging.

jrajahalme replied:
"Envoy rebase is needed to get support for the new '--log-format' command
line option."

rlenglet replied:
Yep, my bad.

rlenglet reviewed (commented):
My bad, they are related. :)

rlenglet commented on pkg/envoy/envoy.go:
```diff
@@ -150,18 +151,35 @@ func StartEnvoy(adminPort uint32, stateDir, logPath string, baseID uint64) *Envo
+				cmd.Stderr = logger
+				cmd.Stdout = logger
+			} else {
+				cmd.Stderr = os.Stderr
```
This is not correct. We can't assume we're logging into stdout / stderr. We have support for Fluentd, etc. in Cilium and we can't break that. And we want to make sure that we blend well with the other logs from Cilium. So those logs must go through logrus.

The right way to do this is to use logrus's `Writer` support:
```
logger := log.WithFields(...).WriterLevel(logrus.DebugLevel)
...
cmd.Stderr = logger
cmd.Stdout = logger
```

That's not great, because we'll want to set the logrus log level according to the Envoy log level. Here even non-debug logs will be logged at Debug level.
But that's a good start.

It may be nice to have different log levels for stdout and stderr:
```
stderrLogger := log.WithFields(...).WriterLevel(logrus.WarnLevel)
stdoutLogger := log.WithFields(...).WriterLevel(logrus.DebugLevel)
...
cmd.Stderr = stderrLogger
cmd.Stdout = stdoutLogger
```

jrajahalme replied:
Thanks for the pointers, will start with this and see if I can get the the log levels from Envoy to logrus.

rlenglet commented on pkg/envoy/envoy.go:
```diff
@@ -150,18 +151,35 @@ func StartEnvoy(adminPort uint32, stateDir, logPath string, baseID uint64) *Envo
-			Compress:   true, // disabled by default
+		var logger *lumberjack.Logger
+		var logFormat string
+		if logPath != "" {
```
I don't this should be one or the other.
I think we should tee the logs from Envoy, and at least always write into the Cilium logs.

jrajahalme committed:
daemon: Merge Envoy logs with cilium logs by default.

Use a file logger for Envoy only if non-empty '--envoy-log' command
line parameter is given. Otherwise merge Envoy logs with Cilium logs
with 'subsys=envoy-<part>', where '<part>' is Envoy's logger name
(e.g., "filter", "upstream", "router", etc.).

When merging Envoy logs to Cilium logs all Envoy log messages will be
logged at "Debug" level. This causes Envoy logs only appear when
Cilium is in debug mode.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

rlenglet commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -33,7 +33,7 @@ cilium-agent
       --enable-policy string              Enable policy enforcement (default "default")
       --enable-tracing                    Enable tracing while determining policy (debugging)
-      --envoy-log string                  Path to Envoy log (default "/var/log/cilium-envoy.log")
+      --envoy-log string                  Path to a separete Envoy log file, if any
```
separete -> separate

ianvernon replied:
Reviewing for docs: please fix this typo.

jrajahalme replied:
Fixed.

rlenglet commented on pkg/envoy/envoy.go:
```diff
@@ -150,18 +150,34 @@ func StartEnvoy(adminPort uint32, stateDir, logPath string, baseID uint64) *Envo
+			// The logs will be reported as coming from the cilium-agent, so
+			// we add the thread id to be able to differentiate between Envoy's
+			// main and worker threads.
+			logFormat = "level=%l msg=\"%v\" subsys=envoy-%n thread=%t"
```
Can you confirm that %v properly quotes the string, esp. transforms `“` into `\”`?

jrajahalme replied:
It does not, e.g.:
```
[2018-03-16 19:54:15.118][30623][debug][config] bazel-out/k8-fastbuild/bin/external/envoy/source/common/config/_virtual_includes/grpc_mux_subscription_lib/common/config/grpc_mux_subscription_impl.h:59] gRPC config for type.googleapis.com/cilium.NetworkPolicy accepted with 10 resources: [name: "f00d::a0b:0:0:dc06"
policy: 43065
...
, name: "f00d::a0b:0:0:367d"
policy: 21941
]
```

rlenglet reviewed (approved):
LGTM. Found one typo and have one question.

ianvernon reviewed (approved):
Approving for docs.
